llvm/lib/Target/X86/X86InstrInfo.cpp

```diff
@@ -92,6 +92,37 @@ X86InstrInfo::X86InstrInfo(X86Subtarget &STI)
       Subtarget(STI), RI(STI.getTargetTriple()) {
 }
 
+const TargetRegisterClass *
+X86InstrInfo::getRegClass(const MCInstrDesc &MCID, unsigned OpNum,
+                          const TargetRegisterInfo *TRI,
+                          const MachineFunction &MF) const {
+  auto *RC = TargetInstrInfo::getRegClass(MCID, OpNum, TRI, MF);
+  // If the target does not have egpr, then r16-r31 will be resereved for all
+  // instructions.
+  if (!RC || !Subtarget.hasEGPR())
+    return RC;
+
+  if (X86II::canUseApxExtendedReg(MCID))
+    return RC;
+
+  switch (RC->getID()) {
+  default:
```

Comment:
Is this optimism for other cases? I think we have other classes like `GR8_ABCD_LRegClassID`, `GR16_ABCDRegClassID` etc.

Reply:
`GR*ABCD`s are not affected at all b/c they only includes `RAX, RCX, RDX, RBX`

Reply:
OK, how about `LOW32_ADDR_ACCESSRegClass`, `LOW32_ADDR_ACCESS_RBPRegClass`?

Reply:
Not affected either. We only need to care about the regclass of operands of instructions. Both of classes are used to check if a reg is the lower 32-bit of GPR only.